File: frontend/src/types/article.ts
export type ArticleStatus = 
  | 'pending' 
  | 'approved' 
  | 'rejected' 
  | 'scheduled' 
  | 'published' 
  | 'failed' 
  | 'duplicate' 
  | 'needs_review';

export interface ImageInfo {
  url: string;
  alt?: string;
  width?: number;
  height?: number;
  is_main: boolean;
}

export interface QualityFactors {
  has_image: number;
  text_length: number;
  freshness: number;
  source_trust: number;
  uniqueness: number;
}

export interface PriorityFactors {
  category_weight: number;
  quality_bonus: number;
  freshness_bonus: number;
  source_bonus: number;
  media_bonus: number;
}

export interface Article {
  id: string;
  token: string;
  project_id?: string;
  source_id?: string;
  source_name?: string;
  url: string;
  title: string;
  description?: string;
  content_clean?: string;
  pub_date?: string;
  category?: string;
  tags?: string[];
  images?: ImageInfo[];
  main_image_url?: string;
  quality_score: number;
  quality_factors?: QualityFactors;
  priority_score: number;
  priority_factors?: PriorityFactors;
  status: ArticleStatus;
  batch_id?: string;
  ai_used: boolean;
  ai_metadata?: Record<string, any>;
  moderated_at?: string;
  moderator_id?: string;
  moderator_name?: string;
  rejection_reason?: string;
  published_at?: string;
  scheduled_at?: string;
  published_target_id?: string;
  published_external_id?: string;
  similar_to_id?: string;
  similarity_score?: number;
  created_at: string;
  updated_at: string;
}

export interface ArticleListItem {
  id: string;
  token: string;
  title: string;
  category?: string;
  source_id?: string;
  source_name?: string;
  status: ArticleStatus;
  quality_score: number;
  priority_score: number;
  has_image: boolean;
  ai_used: boolean;
  pub_date?: string;
  created_at: string;
}

export interface ArticleDetail extends Article {
  similar_articles?: ArticleListItem[];
  versions_count: number;
}

export interface ArticleVersion {
  id: string;
  article_id: string;
  version_number: number;
  title: string;
  content_clean?: string;
  category?: string;
  tags?: string[];
  main_image_url?: string;
  created_at: string;
  created_by_id?: string;
  created_by_name?: string;
  change_summary?: string;
}

export interface ArticleFilters {
  status?: ArticleStatus[];
  category?: string;
  source_id?: string;
  has_image?: boolean;
  ai_used?: boolean;
  min_quality?: number;
  min_priority?: number;
  date_from?: string;
  date_to?: string;
  search?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface ArticleUpdate {
  title?: string;
  description?: string;
  content_clean?: string;
  category?: string;
  tags?: string[];
  main_image_url?: string;
}

export interface ArticleApproveRequest {
  target_id?: string;
  schedule_at?: string;
  use_ai_rewrite?: boolean;
}

export interface ArticlePreview {
  text: string;
  has_image: boolean;
  image_url?: string;
  estimated_length: number;
  warnings: string[];
  is_valid_for_telegram: boolean;
}

File: backend/app/schemas/article.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime

from app.models.article import ArticleStatus


class ImageInfo(BaseModel):
    url: str
    alt: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None
    is_main: bool = False


class QualityFactors(BaseModel):
    has_image: float = 0.0
    text_length: float = 0.0
    freshness: float = 0.0
    source_trust: float = 0.0
    uniqueness: float = 0.0


class PriorityFactors(BaseModel):
    category_weight: int = 0
    quality_bonus: int = 0
    freshness_bonus: int = 0
    source_bonus: int = 0
    media_bonus: int = 0


class ArticleBase(BaseModel):
    title: str = Field(..., min_length=1)
    description: Optional[str] = None
    content_clean: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = None
    main_image_url: Optional[str] = None


class ArticleCreate(ArticleBase):
    url: str
    source_id: Optional[UUID] = None
    content_raw: Optional[str] = None
    pub_date: Optional[datetime] = None


class ArticleUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1)
    description: Optional[str] = None
    content_clean: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = None
    main_image_url: Optional[str] = None


class ArticleResponse(ArticleBase):
    id: UUID
    token: str
    project_id: Optional[UUID] = None
    source_id: Optional[UUID] = None
    url: str
    pub_date: Optional[datetime] = None
    images: Optional[List[ImageInfo]] = None
    quality_score: float
    quality_factors: Optional[QualityFactors] = None
    priority_score: int
    priority_factors: Optional[PriorityFactors] = None
    status: ArticleStatus
    batch_id: Optional[UUID] = None
    ai_used: bool
    ai_metadata: Optional[Dict[str, Any]] = None
    moderated_at: Optional[datetime] = None
    moderator_id: Optional[UUID] = None
    rejection_reason: Optional[str] = None
    published_at: Optional[datetime] = None
    scheduled_at: Optional[datetime] = None
    published_target_id: Optional[UUID] = None
    published_external_id: Optional[str] = None
    similar_to_id: Optional[UUID] = None
    similarity_score: Optional[float] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ArticleListResponse(BaseModel):
    id: UUID
    token: str
    title: str
    category: Optional[str] = None
    source_id: Optional[UUID] = None
    source_name: Optional[str] = None
    status: ArticleStatus
    quality_score: float
    priority_score: int
    has_image: bool
    ai_used: bool
    pub_date: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ArticleDetailResponse(ArticleResponse):
    source_name: Optional[str] = None
    moderator_name: Optional[str] = None
    similar_articles: Optional[List["ArticleListResponse"]] = None
    versions_count: int = 0


class ArticleApproveRequest(BaseModel):
    target_id: Optional[UUID] = None
    schedule_at: Optional[datetime] = None
    use_ai_rewrite: bool = False


class ArticleRejectRequest(BaseModel):
    reason: Optional[str] = None


class ArticleScheduleRequest(BaseModel):
    scheduled_at: datetime
    target_id: Optional[UUID] = None


class ArticleBulkActionRequest(BaseModel):
    article_ids: List[UUID] = Field(..., min_items=1, max_items=50)


class ArticleBulkActionResponse(BaseModel):
    success_count: int
    failed_count: int
    failed_ids: List[UUID] = []
    message: str


class ArticleVersionResponse(BaseModel):
    id: UUID
    article_id: UUID
    version_number: int
    title: str
    content_clean: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = None
    main_image_url: Optional[str] = None
    created_at: datetime
    created_by_id: Optional[UUID] = None
    created_by_name: Optional[str] = None
    change_summary: Optional[str] = None

    class Config:
        from_attributes = True


class ArticlePreviewRequest(BaseModel):
    template_id: Optional[UUID] = None
    target_id: Optional[UUID] = None


class ArticlePreviewResponse(BaseModel):
    text: str
    has_image: bool
    image_url: Optional[str] = None
    estimated_length: int
    warnings: List[str] = []


class ArticleFilterParams(BaseModel):
    status: Optional[List[ArticleStatus]] = None
    category: Optional[str] = None
    source_id: Optional[UUID] = None
    has_image: Optional[bool] = None
    ai_used: Optional[bool] = None
    min_quality: Optional[float] = None
    max_quality: Optional[float] = None
    min_priority: Optional[int] = None
    date_from: Optional[datetime] = None
    date_to: Optional[datetime] = None
    search: Optional[str] = None
    sort_by: str = "created_at"
    sort_order: str = "desc"
